File: kmp/shared/engine/src/commonMain/kotlin/game/vinto/engine/Replay.kt
package game.vinto.engine

import game.vinto.shapes.GameAction
import game.vinto.shapes.GameRecording
import game.vinto.shapes.GameState
import game.vinto.shapes.assertRecordingVersion
import game.vinto.shapes.canonicalizeGameState
import game.vinto.shapes.hashGameState

/**
 * Replays a `GameRecording` through the engine — the cross-implementation parity harness.
 *
 * The Kotlin engine runs the same recordings TypeScript wrote and must reach a byte-identical
 * canonical state after *every* action. A divergence report names the first action where the
 * two disagree, which localises a porting mistake to a single handler. That per-action
 * granularity is the point: comparing only final states would tell you the engine is wrong
 * without telling you where.
 *
 * Replay reconstructs *engine* state only. `turnActions`/`roundActions` are written by the
 * client, not the reducer, so a replayed state legitimately has no history — which is why
 * comparison is by canonical hash, and why the canonical form excludes history.
 *
 * Ported from `packages/engine/src/lib/replay.ts`, with one addition: `HANDLER_UNPORTED`,
 * which exists only while phase 4 is in progress.
 */
enum class DivergenceReason {
    ACTION_REJECTED,
    HASH_MISMATCH,
    FINAL_STATE_MISMATCH,

    /** Temporary: the port has not reached this handler yet. */
    HANDLER_UNPORTED,
}

data class ReplayDivergence(
    /** Index into `recording.actions`; -1 for a final-state mismatch. */
    val index: Int,
    val reason: DivergenceReason,
    val action: GameAction? = null,
    val detail: String,
    val expectedHash: String? = null,
    val actualHash: String? = null,
)

data class ReplayResult(
    val ok: Boolean,
    /** Number of actions successfully applied. */
    val steps: Int,
    val finalState: GameState,
    val divergence: ReplayDivergence? = null,
)

/**
 * @param verifyFinalState off when replaying a recording whose `finalState` was captured by
 *   a different implementation's client.
 */
fun replayRecording(recording: GameRecording, verifyFinalState: Boolean = true): ReplayResult {
    // Throws on an unknown version rather than producing a confusing divergence.
    assertRecordingVersion(recording)

    var state = recording.initialState

    for ((index, entry) in recording.actions.withIndex()) {
        val result = try {
            GameEngine.reduce(state, entry.action)
        } catch (unported: UnportedHandlerException) {
            return ReplayResult(
                ok = false,
                steps = index,
                finalState = state,
                divergence = ReplayDivergence(
                    index = index,
                    reason = DivergenceReason.HANDLER_UNPORTED,
                    action = entry.action,
                    detail = "No Kotlin handler for ${unported.actionType} yet",
                ),
            )
        }

        if (result is ReduceResult.Failure) {
            return ReplayResult(
                ok = false,
                steps = index,
                finalState = state,
                divergence = ReplayDivergence(
                    index = index,
                    reason = DivergenceReason.ACTION_REJECTED,
                    action = entry.action,
                    detail = "Engine rejected ${entry.action.type}: ${result.reason}",
                ),
            )
        }

        state = result.state

        val expected = entry.stateHash
        if (expected != null) {
            val actual = hashGameState(state)
            if (actual != expected) {
                return ReplayResult(
                    ok = false,
                    steps = index + 1,
                    finalState = state,
                    divergence = ReplayDivergence(
                        index = index,
                        reason = DivergenceReason.HASH_MISMATCH,
                        action = entry.action,
                        detail = "State after ${entry.action.type} does not match the recorded hash",
                        expectedHash = expected,
                        actualHash = actual,
                    ),
                )
            }
        }
    }

    if (verifyFinalState) {
        val expected = hashGameState(recording.finalState)
        val actual = hashGameState(state)
        if (expected != actual) {
            return ReplayResult(
                ok = false,
                steps = recording.actions.size,
                finalState = state,
                divergence = ReplayDivergence(
                    index = -1,
                    reason = DivergenceReason.FINAL_STATE_MISMATCH,
                    detail = "Replayed final state does not match the recorded final state",
                    expectedHash = expected,
                    actualHash = actual,
                ),
            )
        }
    }

    return ReplayResult(ok = true, steps = recording.actions.size, finalState = state)
}

/** Human-readable divergence report for CI logs. */
fun formatDivergence(divergence: ReplayDivergence, stateAtDivergence: GameState? = null): String =
    buildString {
        appendLine("Divergence at action index ${divergence.index} (${divergence.reason})")
        appendLine("  ${divergence.detail}")
        divergence.action?.let { appendLine("  action: ${it.type}") }
        divergence.expectedHash?.let {
            appendLine("  expected: $it")
            appendLine("  actual:   ${divergence.actualHash}")
        }
        stateAtDivergence?.let { appendLine("  state: ${canonicalizeGameState(it)}") }
    }

File: legacy-web/packages/engine/src/lib/replay.ts
import {
  GameAction,
  GameRecording,
  GameState,
  assertRecordingVersion,
  canonicalizeGameState,
  hashGameState,
  rehydrateGameState,
} from '@vinto/shapes';
import { GameEngine } from './game-engine';

/**
 * Replays a `GameRecording` through the engine.
 *
 * This is the cross-implementation parity harness: the Kotlin engine runs the same
 * recordings and must reach byte-identical canonical states after every action. A
 * divergence report names the first action where the two disagree, which localises a
 * porting mistake to a single handler.
 *
 * Note that replay reconstructs *engine* state only. `turnActions`/`roundActions` are
 * written by the client, not by the reducer, so a replayed state legitimately has no
 * history. Comparison is therefore by canonical hash, which excludes history by design.
 */

export type ReplayDivergenceReason =
  | 'action-rejected'
  | 'hash-mismatch'
  | 'final-state-mismatch';

export interface ReplayDivergence {
  /** Index into `recording.actions`; -1 for a final-state mismatch. */
  index: number;
  reason: ReplayDivergenceReason;
  action?: GameAction;
  detail: string;
  expectedHash?: string;
  actualHash?: string;
}

export interface ReplayResult {
  ok: boolean;
  /** Number of actions successfully applied. */
  steps: number;
  finalState: GameState;
  /** State after each applied action; used to fill hashes when writing fixtures. */
  states: GameState[];
  divergence?: ReplayDivergence;
}

export interface ReplayOptions {
  /**
   * Verify `finalState` against the recording. Off when replaying a recording whose
   * `finalState` was captured by a different implementation's client.
   */
  verifyFinalState?: boolean;
}

export async function replayRecording(
  recording: GameRecording,
  options: ReplayOptions = {},
): Promise<ReplayResult> {
  const { verifyFinalState = true } = options;

  // Throws on an unknown version rather than producing a confusing divergence.
  assertRecordingVersion(recording);

  let state = rehydrateGameState(recording.initialState);
  const states: GameState[] = [];

  for (let index = 0; index < recording.actions.length; index++) {
    const entry = recording.actions[index];
    const result = GameEngine.reduce(state, entry.action);

    if (!result.success) {
      return {
        ok: false,
        steps: index,
        finalState: state,
        states,
        divergence: {
          index,
          reason: 'action-rejected',
          action: entry.action,
          detail: `Engine rejected ${entry.action.type}: ${result.reason}`,
        },
      };
    }

    state = result.state;
    states.push(state);

    if (entry.stateHash) {
      const actualHash = await hashGameState(state);
      if (actualHash !== entry.stateHash) {
        return {
          ok: false,
          steps: index + 1,
          finalState: state,
          states,
          divergence: {
            index,
            reason: 'hash-mismatch',
            action: entry.action,
            detail: `State after ${entry.action.type} does not match the recorded hash`,
            expectedHash: entry.stateHash,
            actualHash,
          },
        };
      }
    }
  }

  if (verifyFinalState && recording.finalState) {
    const expectedHash = await hashGameState(
      rehydrateGameState(recording.finalState),
    );
    const actualHash = await hashGameState(state);

    if (expectedHash !== actualHash) {
      return {
        ok: false,
        steps: recording.actions.length,
        finalState: state,
        states,
        divergence: {
          index: -1,
          reason: 'final-state-mismatch',
          detail:
            'Replayed final state does not match the recorded final state',
          expectedHash,
          actualHash,
        },
      };
    }
  }

  return {
    ok: true,
    steps: recording.actions.length,
    finalState: state,
    states,
  };
}

/** Human-readable divergence report for the CLI and CI logs. */
export function formatDivergence(
  divergence: ReplayDivergence,
  finalState?: GameState,
): string {
  const lines = [
    `Divergence at action index ${divergence.index} (${divergence.reason})`,
    `  ${divergence.detail}`,
  ];

  if (divergence.action) {
    lines.push(`  action: ${JSON.stringify(divergence.action)}`);
  }
  if (divergence.expectedHash) {
    lines.push(`  expected: ${divergence.expectedHash}`);
    lines.push(`  actual:   ${divergence.actualHash}`);
  }
  if (finalState) {
    lines.push(`  state at divergence: ${canonicalizeGameState(finalState)}`);
  }

  return lines.join('\n');
}
